pkg/manager/member/tiflash_scaler.go

```diff
@@ -50,113 +52,157 @@ func (s *tiflashScaler) Scale(meta metav1.Object, oldSet *apps.StatefulSet, newS
+	scaleOutOne := func(ordinal int32) error {
+		_, err := s.deleteDeferDeletingPVC(tc, v1alpha1.TiFlashMemberType, ordinal)
+		if err != nil {
+			return err
+		}
+		return nil
 	}
 
-	setReplicasAndDeleteSlots(newSet, replicas, deleteSlots)
-	return nil
+	var (
+		errs             []error
+		finishedOrdinals = sets.NewInt32()
+	)
+	for _, ordinal := range ordinals {
+		err := scaleOutOne(ordinal)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			finishedOrdinals.Insert(ordinal)
+		}
+	}
+	setReplicasAndDeleteSlotsByFinished(1, newSet, oldSet, ordinals, finishedOrdinals)
+	return errorutils.NewAggregate(errs)
 }
 
 func (s *tiflashScaler) ScaleIn(meta metav1.Object, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	ns := meta.GetNamespace()
+	tcName := meta.GetName()
 	tc, ok := meta.(*v1alpha1.TidbCluster)
 	if !ok {
+		klog.Errorf("tiflashScaler.ScaleIn: failed to convert cluster %s/%s, scale in will do nothing", meta.GetNamespace(), meta.GetName())
 		return nil
 	}
 
-	ns := tc.GetNamespace()
-	tcName := tc.GetName()
-	// we can only remove one member at a time when scaling in
-	_, ordinal, replicas, deleteSlots := scaleOne(oldSet, newSet)
-	resetReplicas(newSet, oldSet)
-
-	klog.Infof("scaling in tiflash statefulset %s/%s, ordinal: %d (replicas: %d, delete slots: %v)", oldSet.Namespace, oldSet.Name, ordinal, replicas, deleteSlots.List())
-	// We need delete store from cluster before decreasing the statefulset replicas
-	podName := ordinalPodName(v1alpha1.TiFlashMemberType, tcName, ordinal)
-	pod, err := s.deps.PodLister.Pods(ns).Get(podName)
-	if err != nil {
-		return fmt.Errorf("tiflashScaler.ScaleIn: failed to get pods %s for cluster %s/%s, error: %s", podName, ns, tcName, err)
+	scaleInParallelism := 1
+	if tc.Spec.TiFlash.ScaleInParallelism != nil {
+		scaleInParallelism = int(*tc.Spec.TiFlash.ScaleInParallelism)
 	}
 
-	// TODO: Update Webhook to support TiFlash
-	// if controller.PodWebhookEnabled {
-	// 	setReplicasAndDeleteSlots(newSet, replicas, deleteSlots)
-	// 	return nil
-	// }
+	_, ordinals, replicas, deleteSlots := scaleMulti(oldSet, newSet, scaleInParallelism)
 
-	for _, store := range tc.Status.TiFlash.Stores {
-		if store.PodName == podName {
-			state := store.State
-			id, err := strconv.ParseUint(store.ID, 10, 64)
-			if err != nil {
-				return err
-			}
-			if state != v1alpha1.TiKVStateOffline {
-				if err := controller.GetPDClient(s.deps.PDControl, tc).DeleteStore(id); err != nil {
-					klog.Errorf("tiflash scale in: failed to delete store %d, %v", id, err)
-					return err
+	klog.Infof("scaling in tiflash statefulset %s/%s, ordinal: %v (replicas: %d, delete slots: %v), scaleInParallelism: %v", oldSet.Namespace, oldSet.Name, ordinals, replicas, deleteSlots.List(), scaleInParallelism)
+
+	scaleInOne := func(ordinal int32) (bool, error) {
```

Comment:
It seems only return `error` is enough? currently, always `false + non-nil error` and `true + nil error` returned

Reply:
Emmm, when call `preCheckUpStores` it may return `false + nil error` so maybe we should keep them both.

A little tricky here but I think keep the same logic with former version, any suggestion to refine it?

Reply:
I refactor the code according to previous comment, maybe it would be more clear a little bit.

Reply:
@duduainankai Where is `preCheckUpStores` called now?
And where is `false + nil` returned?

Reply:
@DanielZhangQD 

`preCheckUpStores` has been refactored to return a `bool` only, move the `pdClient.GetStores()` outside. 

But `scaleInOne` may still return a `false + nil` result if `preCheckUpStores` return a false. Try to keep the same logic with before. 

But I see you comment below that we can return an error here, so should we treat `preCheckUpStores` not pass as an error and refactor to `scaleInOne := func(ordinal int32) error` ?

Reply:
Oh sorry, I thought it's in `tikv_scaler.go`.